volume binder: simplify API

pohly opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:

Explicitly listing the reasons why the volume binder rejected a node via boolean return values is not going to scale once the number of reasons grows, which it might at some point (but not today...).

**Special notes for your reviewer**:

This was originally submitted in https://github.com/kubernetes/kubernetes/pull/88160 together with adding one more failure reason, but that additional check missed some aspects and can't be merged. This refactoring was deemed worthwhile to merge, hence this PR.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

pohly committed:
volume scheduler: move reason strings into volume code

The scheduler doesn't really need to know in detail which reasons
rendered a node unusable for a node. All it needs from the volume
binder is a list of reasons that it then can present to the user.

This seems a bit cleaner. But the main reason for the change is that
it simplifies the checking of CSI inline volumes and perhaps later
capacity checking. Both will lead to new failure reasons, which then
can be added without changing the interface.

pohly commented:
/assign @ahg-g 
/assign @msau42

msau42 commented on pkg/controller/volume/scheduling/scheduler_binder.go:
```diff
@@ -78,16 +85,16 @@ type InTreeToCSITranslator interface {
 //    while the actual binding operation occurs in the background.
 type SchedulerVolumeBinder interface {
-	// FindPodVolumes checks if all of a Pod's PVCs can be satisfied by the node.
+	// FindPodVolumes checks if all of a Pod's volumes can be satisfied by the node.
```
change comment back to PVC since that's the current behavior.

pohly replied:
Okay.

msau42 commented on pkg/controller/volume/scheduling/scheduler_binder.go:
```diff
@@ -78,16 +85,16 @@ type InTreeToCSITranslator interface {
 	//
 	// This function is called by the volume binding scheduler predicate and can be called in parallel
-	FindPodVolumes(pod *v1.Pod, node *v1.Node) (unboundVolumesSatisified, boundVolumesSatisfied bool, err error)
+	FindPodVolumes(pod *v1.Pod, node *v1.Node) (reasons []string, err error)
```
should we make the reasons typed so that we can easily search and find the reasons?

pohly replied:
The reason might also be the result of a `fmt.Sprintf`, it doesn't necessarily have to be a constant. Such a string could of course also be typecast - okay, let's introduce a type.

pohly replied:
Done. This had implications throughout the code. I've kept it in separate commit, so it is possible to compare the version with and without special types in the PR.

@msau42 what do you prefer?

msau42 commented on pkg/controller/volume/scheduling/scheduler_binder.go:
```diff
@@ -238,7 +245,7 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, node *v1.Node) (unboundVolume
 				if selectedNode != node.Name {
 					// Fast path, skip unmatched node
-					return false, boundVolumesSatisfied, nil
+					return []string{ErrReasonBindConflict}, nil
```
boundVolumesSatisfied could also be false at this point.

pohly replied:
I know (https://github.com/kubernetes/kubernetes/pull/88160#discussion_r379496960). I decided to keep it simple here because in the end all that matters is that *some* reason is returned. I can add an if check for `boundVolumesSatisfied` here if you think it is important.

msau42 replied:
I think it'll be nice to return all reasons if we know it instead of having it depend on ordering of checks

pohly replied:
If we allow the use of `goto` here, then it can be done without code duplication. Changed.

pohly committed:
volume scheduler: introduce special string type

This makes it possible to search for the special strings more easily
(https://github.com/kubernetes/kubernetes/pull/88230#discussion_r382367043).

pohly committed:
bazel update

pohly commented:
The test failures seem unrelated to this PR (https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/88230/pull-kubernetes-verify/1232253416395247616, https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/88230/pull-kubernetes-bazel-test/1232253414742691841).

msau42 commented on pkg/controller/volume/scheduling/scheduler_binder.go:
```diff
@@ -78,16 +96,16 @@ type InTreeToCSITranslator interface {
 //    while the actual binding operation occurs in the background.
 type SchedulerVolumeBinder interface {
-	// FindPodVolumes checks if all of a Pod's PVCs can be satisfied by the node.
+	// FindPodVolumes checks if all of a Pod's PVC can be satisfied by the node.
```
PVCs

pohly replied:
Oh, bummer. Sorry, fixed.

FWIW, my motivation for the change to "volumes" was that this is the interface documentation, not the documentation for the current implementation of that interface. As far as the caller is concerned, it shouldn't matter whether the implementation checks PVCs or also something else. With this change of booleans with fixed meanings to a slice of strings the caller really doesn't need to care anymore, so it seemed like a good time to replace it.

But I'm also fine with documentation the full behavior here.

msau42 commented on pkg/controller/volume/scheduling/scheduler_binder.go:
```diff
@@ -237,8 +255,9 @@ func (b *volumeBinder) FindPodVolumes(pod *v1.Pod, node *v1.Node) (unboundVolume
-					return false, boundVolumesSatisfied, nil
+					// Fast path, skip unmatched node.
+					unboundVolumesSatisfied = false
+					goto done
```
I don't see goto used anywhere substantial in our codebase. How about using defer, and changing this back to return?

pohly replied:
So a `return nil, nil` here instead of `goto done` and then in a deferred function check if `err == nil` and if yes, then add the reason strings? I had considered that, but it seemed more complicated and I wasn't sure which approach would be considered better style. I can change it.

pohly replied:
Or better, just plain `return` - then it is obvious that return value handling happens elsewhere.

pohly commented:
Updated and rebased commits (there was a concurrent spelling fix in the binder test).

pohly commented:
/retest

pohly commented:
/retest

pohly commented:
/retest

msau42 commented:
/lgtm
/approve
Thanks!

pohly commented:
"Yet another rebase" (TM), this time because of a conflict with 18ffaf560843e764ef47c147a59b158d0baffb17.

pohly commented:
/retest

pohly commented:
@ahg-g : can you approve for sig-scheduling?

ahg-g commented:
/approve

msau42 commented:
/lgtm
